Code/C++/avl_tree.cpp

```diff
@@ -0,0 +1,429 @@
+    p3->left = p2;                          //C->left = B
+
+    return p3;                              //C is the new root
+}
+
+//right-left rotation function
+node* RLrotation(node *root)
+{
+    node *p1;
+    node *p2;
+    node *p3;
+
+    p1 = root;                              //A
+    p2 = root->right;                       //B
+    p3 = root->right->left;                 //C
+
+    p1->right = p3->left;                   //A->right = C->left
+    p2->left = p3->right;                   //B->left = C->right
+    p3->right = p2;                         //C->right = B
+    p3->left = p1;                          //C->left =
+
+    return p3;                              //C is the new root
+}
+
+//return the height of a particular node
+int nodeheight(node* root)
+{
+    if(root && root->left)
+    {
+        if(root->left->height > root->right->height)
+            return (root->left->height + 1);
+    }
+    else if(root && root->right)
+    {
+        if(root->right->height > root->left->height)
+            return (root->right->height + 1);
+    }
+}
+
+//returns the balance factor of a node
+int BalanceFactor(node* root)
+{
+    if(root == NULL)
+        return 0;
+    else
+        return(height(root->left) - height(root->right));
+}
+
+//creates a new node of the given data
+node* create(int data)
+{
+    node *newnode = new node;
+    newnode->info = data;
+    newnode->left = NULL;
+    newnode->right = NULL;
+    newnode->height = 1;
+    return newnode;
+}
+
+//inserts a node in the AVL tree
+node* AVLinsertion(node* root , int data)
+{
+    if(root == NULL)
+    {
+        root = create(data);
+        return root;
+    }
+
+    else if(data < root->info)
+        root->left = AVLinsertion(root->left , data);
+
+    else
+        root->right = AVLinsertion(root->right , data);
+
+    //node's height will be max of (l-height,r-height) + 1
+    root->height = (max(height(root->left) , height(root->right))+1);
+
+    //Insertion is done now conditions for rotation if needed.
+    if(BalanceFactor(root)==2 && BalanceFactor(root->left)==1)
+        return LLrotation(root);
+
+    if(BalanceFactor(root)==2 && BalanceFactor(root->left)==-1)
+        return LRrotation(root);
+
+    if(BalanceFactor(root)==-2 && BalanceFactor(root->right)==-1)
+        return RRrotation(root);
+
+    if(BalanceFactor(root)==-2 && BalanceFactor(root->right)==1)
+        return RLrotation(root);
+
+    return root;
+}
+
+//return the minimum node of the tree
+node* minimumnode(node *root)
+{
+    if(root == NULL)
+        return NULL;
+
+    if(root->left == NULL)
+        return root;
+
+    else
+        minimumnode(root->left);
+}
+
+//return the maximum node of the tree
+node* maximumnode(node *root)
+{
+    if(root == NULL)
+        return NULL;
+
+    if(root->right == NULL)
+        return root;
+
+    else
+        maximumnode(root->right);
+}
+
+//deletes an element from tree
+node* deleteelement(node *root , int data)
```

Comment:
it would be better to name it as delete_element

Reply:
done sir, please review